### PR TITLE
DRAFT: Initial Support for LL-HLS 

### PR DESCRIPTION
**Pull request by JuanMorenoS**
#### Summary
This pull request introduces initial support for Low-Latency HLS (LL-HLS). The following changes have been made to start accommodating the LL-HLS specification:

#### Changes Implemented
1. **New LL-HLS Tags Support:**
   - Added parsers for LL-HLS specific tags, including `#EXT-X-SERVER-CONTROL`, `#EXT-X-PART-INF`, `#EXT-X-PART`, `#EXT-X-SKIP`, `#EXT-X-PRELOAD-HINT`, and `#EXT-X-RENDITION-REPORT`.

2. **Updated `MediaPlaylistTag` Enum:**
   - Extended the `MediaPlaylistTag` enum to include variants for the new LL-HLS tags.

3. **Enhanced `media_playlist_from_tags` Function:**
   - Updated the `media_playlist_from_tags` function to properly handle and integrate the new LL-HLS tags into the `MediaPlaylist` structure.
   - Ensured `Part` segments are correctly added to the `parts` field of `MediaSegment`.

4. **Parser Integrations:**
   - Integrated the new parsers into the `media_playlist_tag` function to recognize and process the LL-HLS tags during playlist parsing.

5. **Test Additions and Adjustments:**
   - Added tests to verify the correct parsing and serialization of LL-HLS tags.
   - Adjusted existing tests to ensure compatibility with the new LL-HLS features.

#### Next Steps
- Continue to enhance LL-HLS support by adding more detailed validation and error handling for LL-HLS specific scenarios.
- Optimize performance and ensure full compliance with the latest LL-HLS specifications.

### CODE REVIEW

File: src/playlist.rs [hunk 1]
```diff
@@ -1144,6 +1208,222 @@ impl DateRange {
...
+        let part_target = unquoted_string_parse!(attrs, "PART-TARGET", |s: &str| s
+            .parse::<f64>()
+            .map_err(|err| format!("Failed to parse PART-TARGET attribute: {}", err)))
+        .ok_or_else(|| String::from("EXT-X-PART-INF without mandatory PART-TARGET attribute"))?;
+
+        Ok(PartInf { part_target })
+    }
+
+    pub(crate) fn write_to<T: Write>(&self, w: &mut T) -> std::io::Result<()> {
+        write!(w, "#EXT-X-PART-INF:PART-TARGET={}", self.part_target)?;
```
Comment (rutgersc) [inline src/playlist.rs:1273]: I believe everywhere where a `f64` is written out, it needs to use the same if/else construction as in [this](https://github.com/rutgersc/m3u8-rs/pull/72/files#diff-294648967369087cf3619793bd50043a2b39b5f481523d7c1636cc3e8d572a5bR891-R899) example. Some parsers cannot deal with floats when they don't have decimals, so it's good to apply it to all the new `f64` fields so it's consistent with the rest.

File: src/playlist.rs [hunk 2]
```diff
@@ -1144,6 +1208,222 @@ impl DateRange {
...
+}
+
+impl ServerControl {
+    pub(crate) fn from_hashmap(
+        mut attrs: HashMap<String, QuotedOrUnquoted>,
+    ) -> Result<ServerControl, String> {
+        let can_skip_until = unquoted_string_parse!(attrs, "CAN-SKIP-UNTIL", |s: &str| s
+            .parse::<f64>()
+            .map_err(|err| format!("Failed to parse CAN-SKIP-UNTIL attribute: {}", err)));
+        let hold_back = unquoted_string_parse!(attrs, "HOLD-BACK", |s: &str| s
```
Comment (rutgersc) [inline src/playlist.rs:1227]: I'm also now seeing `CAN-SKIP-DATERANGES` under [4.4.3.8](https://author-tools.ietf.org/iddiff?url1=draft-pantos-hls-rfc8216bis-06&url2=draft-pantos-hls-rfc8216bis-07&difftype=--html) in the draft, could add that one as well

File: src/playlist.rs [hunk 3]
```diff
@@ -1144,6 +1208,222 @@ impl DateRange {
...
+    pub independent: bool,
+    pub gap: bool,
+    pub byte_range: Option<ByteRange>,
+}
+
+impl Part {
+    pub(crate) fn from_hashmap(
+        mut attrs: HashMap<String, QuotedOrUnquoted>,
+    ) -> Result<Part, String> {
+        let uri = quoted_string!(attrs, "URI").unwrap_or_default();
```
Comment (rutgersc) [inline src/playlist.rs:1291]: The uri is mandatory

File: src/playlist.rs [hunk 4]
```diff
@@ -1048,6 +1081,37 @@ impl ByteRange {
...
+            write!(f, "@{}", offset)?;
+        }
+        Ok(())
+    }
+}
+
+impl FromStr for ByteRange {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<ByteRange, String> {
```
Comment (rutgersc) [inline src/playlist.rs:1097]: Doesn't this do the same as the one in [parser.rs](https://github.com/rutgersc/m3u8-rs/blob/master/src/parser.rs#L735)? Maybe these could share the same implementation?

**Review by rutgersc — COMMENTED** [review]
Thanks for contributing! I left a few comments. The one about floats is probably something that is still required for a new version

**Comment by JeWe37** [comment]
Hey, are there any plans to finish this up? If not, would you mind me taking this over, looking at fixing up some of the comments and getting this ready to merge?

**Comment by rutgersc** [comment]
I don't think there are any plans, by the looks of it. Feel free to make the changes.

**Comment by JuanMorenoS** [comment]
Sorry guys, I’ve been a bit busy with work, feel free to make the necessary changes to merge this up :)

**Pull request closed without merging**